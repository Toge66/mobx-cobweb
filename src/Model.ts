/***************************************************
 * Created by nanyuantingfeng on 2020/6/2 12:53. *
 ***************************************************/
import { Model as _Model } from './datx'
import { withStorage } from './mixins/withStorage'
import { withNetActions } from './mixins/withNetActions'

/**** For DTS File (Don`t remove it) ****/
import { INetActionsMixinForModel } from './interfaces/INetActionsMixin'
import { IModelConstructor } from './datx'

export class Model extends withStorage(withNetActions(_Model)) {}
