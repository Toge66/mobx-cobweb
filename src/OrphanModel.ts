/***************************************************
 * Created by nanyuantingfeng on 2020/7/21 11:40. *
 ***************************************************/
import { Model } from './Model'
import { attribute } from './index'

export class OrphanModel extends Model {
  static enableAutoId = false
  static __$orphan_id__ = '__$orphan_id__(0)'
  static preprocess(data: any = {}) {
    data.__$orphan_id__ = OrphanModel.__$orphan_id__
    return data
  }

  @attribute({ isIdentifier: true }) private __$orphan_id__: number
}
