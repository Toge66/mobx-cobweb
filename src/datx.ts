/***************************************************
 * Created by nanyuantingfeng on 2020/7/28 09:28. *
 ***************************************************/
export * from '@issues-beta/datx'
