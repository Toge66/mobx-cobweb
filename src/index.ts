/***************************************************
 * Created by nanyuantingfeng on 2019/11/26 12:22. *
 ***************************************************/
export * from 'datx'
export * from './interfaces'
export * from './implements/helpers/model'

export type { ISkeletonCollection } from './interfaces/ISkeletonCollection'
export type { ISkeletonModel } from './interfaces/ISkeletonModel'
export type { ISkeletonView } from './interfaces/ISkeletonView'
export type { IRawResponse } from './interfaces/IRawResponse'
export type { IRequestOptions } from './interfaces/IRequestOptions'
export type { IResponseView } from './interfaces/IResponseView'

export { ResponseView } from './implements/ResponseView'
export { GenericModel } from './implements/GenericModel'
export { GenericView } from './implements/GenericView'
export { setNetworkAdapter } from './implements/NetworkUtils'
export { clearAllCache, clearCacheByType } from './implements/cache'

export { skeleton } from './mixin'
export * from './views'

export * from './implements/adapter/NetworkAdapter'
