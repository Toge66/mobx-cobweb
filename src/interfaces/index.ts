/***************************************************
 * Created by nanyuantingfeng on 2019/11/26 14:12. *
 ***************************************************/
export * from './types'
export * from './IRawResponse'
export * from './IRequestOptions'
export * from './IRawData'
export * from './IResponseHeaders'
export * from './INetworkAdapter'
