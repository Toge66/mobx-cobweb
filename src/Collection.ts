/***************************************************
 * Created by nanyuantingfeng on 2020/6/2 12:55. *
 ***************************************************/
import { Collection as _Collection, IModelConstructor, IType, PureModel, ICollectionConstructor } from 'datx'
import { withNetPatches } from './mixins/withNetPatches'
import { withStorage } from './mixins/withStorage'
import { INetPatchesCollectionMixin } from './interfaces/INetPatchesCollectionMixin'
import { IStorageCollectionMixin } from './interfaces/IStorageCollectionMixin'

const WithNetPatchesCollection: ICollectionConstructor<
  IStorageCollectionMixin<_Collection> & INetPatchesCollectionMixin<_Collection> & _Collection
> = withStorage(withNetPatches(_Collection))

export class Collection extends WithNetPatchesCollection {
  static register<T extends PureModel>(O: IModelConstructor<T>) {
    if (!this.types.find((Q) => Q.type === O.type)) {
      this.types.push(O)
    }
  }

  register<T extends PureModel>(O: IModelConstructor<T>) {
    ;((this.constructor as unknown) as Collection).register(O)
  }

  findOrphan<T extends PureModel>(model?: IType | IModelConstructor<T>): T {
    return this.findOne(model, '__$orphan_id__(0)')
  }
}
