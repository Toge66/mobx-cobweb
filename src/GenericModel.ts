/***************************************************
 * Created by nanyuantingfeng on 2019/11/26 12:22. *
 ***************************************************/
import { IModelConstructor, PureModel } from 'datx'
import { withNetActions } from './mixins/withNetActions'
import { INetActionsMixin } from './interfaces/INetActionsMixin'
const WithNetActionModel: IModelConstructor<INetActionsMixin<PureModel> & PureModel> = withNetActions(PureModel)

export class GenericModel extends WithNetActionModel {}
