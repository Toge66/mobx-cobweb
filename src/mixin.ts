/***************************************************
 * Created by nanyuantingfeng on 2019/11/26 12:22. *
 ***************************************************/
import {
  ICollectionConstructor,
  IModelConstructor,
  isCollection,
  isModel,
  isView,
  IViewConstructor,
  PureCollection,
  PureModel,
  View
} from 'datx'

import { decorateCollection } from './implements/decorateCollection'
import { decorateModel } from './implements/decorateModel'
import { decorateView } from './implements/decorateView'
import { ISkeletonCollection, ISkeletonModel, ISkeletonModelConstructor } from './interfaces'

export function skeleton<T extends PureModel>(
  BaseClass: IModelConstructor<T>
): ISkeletonModelConstructor<T & ISkeletonModel>

export function skeleton<T extends PureCollection>(
  BaseClass: ICollectionConstructor<T>
): ICollectionConstructor<T & ISkeletonCollection>

export function skeleton<T extends PureModel | PureCollection | View>(
  BaseClass: IModelConstructor<T> | ICollectionConstructor<T> | IViewConstructor<T>
) {
  if (isModel(BaseClass)) {
    return decorateModel(BaseClass as typeof PureModel)
  } else if (isCollection(BaseClass)) {
    return decorateCollection(BaseClass as typeof PureCollection)
  } else if (isView(BaseClass)) {
    return decorateView<T>(BaseClass as typeof View)
  }

  throw new Error('The instance needs to be a model, collection or a view')
}
