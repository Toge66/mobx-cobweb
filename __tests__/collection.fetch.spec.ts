import { collection, useFixtureGetById, useFixturesByGET } from './config'
import { fetchModelRefs, View, modelToJSON, getRefId } from '../src'
import Me from './models/Me'
import Staff from './models/Staff'

describe('collection.fetch', () => {
  let scope: any = null

  beforeEach(() => {
    scope = useFixturesByGET()
    collection.removeAll(Me)
    collection.removeAll(Staff)
  })

  test('should be fetched a Model By Id/Ids', async () => {
    await collection.fetch(Me)
    const me = collection.findAll<Me>(Me)[0]
    expect(me.staff).toBeNull()
    expect((getRefId(me, 'staff') as any).id).toBe('cdb28c900c75')

    useFixtureGetById(Staff.endpoint)
    await collection.fetch(Staff, (getRefId(me, 'staff') as any).id)
    expect(me.staff.id).toEqual('cdb28c900c75')
    expect(modelToJSON(me)).toMatchSnapshot()
    expect(modelToJSON(me.staff)).toMatchSnapshot()
    const meV = new View(Me, collection, null, [me])
    expect(meV.list[0] === me).toBeTruthy()
  })

  test('should be fetched all Refs Models fetchModelRefs', async () => {
    await collection.fetch(Me)
    const me = collection.findAll<Me>(Me)[0]
    expect(me.staff).toBeNull()
    expect((getRefId(me, 'staff') as any).id).toBe('cdb28c900c75')

    useFixtureGetById(Staff.endpoint)
    await fetchModelRefs(me)
    expect(me.staff.id).toEqual('cdb28c900c75')
    expect(modelToJSON(me)).toMatchSnapshot()
    expect(modelToJSON(me.staff)).toMatchSnapshot()
    const meV = new View(Me, collection, null, [me])
    expect(meV.list[0] === me).toBeTruthy()
  })
})
