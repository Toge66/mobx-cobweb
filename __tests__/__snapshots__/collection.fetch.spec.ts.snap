// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collection.fetch should be fetched a Model By Id/Ids 1`] = `
Object {
  "__$orphan_id__": "__$orphan_id__(0)",
  "__META__": Object {
    "MODEL_PERSISTED_FIELD": false,
    "collection": undefined,
    "fields": Object {
      "__$orphan_id__": Object {
        "defaultValue": undefined,
        "referenceDef": false,
      },
      "staff": Object {
        "defaultValue": undefined,
        "referenceDef": Object {
          "model": "organization.Staff",
          "type": 0,
        },
      },
    },
    "id": "__$orphan_id__(0)",
    "type": "organization.Staff.Me",
  },
  "staff": Object {
    "id": "cdb28c900c75",
    "type": "organization.Staff",
  },
}
`;

exports[`collection.fetch should be fetched a Model By Id/Ids 2`] = `
Object {
  "__META__": Object {
    "MODEL_PERSISTED_FIELD": true,
    "collection": undefined,
    "fields": Object {
      "code": Object {
        "defaultValue": undefined,
        "referenceDef": false,
      },
      "defaultDepartment": Object {
        "defaultValue": undefined,
        "referenceDef": Object {
          "model": "organization.Department",
          "type": 0,
        },
      },
      "departments": Object {
        "defaultValue": undefined,
        "referenceDef": Object {
          "model": "organization.Department",
          "property": undefined,
          "type": 1,
        },
      },
      "id": Object {
        "defaultValue": undefined,
        "referenceDef": false,
      },
      "name": Object {
        "defaultValue": undefined,
        "referenceDef": false,
      },
    },
    "id": "cdb28c900c75",
    "type": "organization.Staff",
  },
  "code": 41738,
  "defaultDepartment": Object {
    "id": "66f58b59c384",
    "type": "organization.Department",
  },
  "departments": Array [
    Object {
      "id": "66f58b59c384",
      "type": "organization.Department",
    },
  ],
  "id": "cdb28c900c75",
  "name": "Principal",
}
`;

exports[`collection.fetch should be fetched all Refs Models fetchModelRefs 1`] = `
Object {
  "__$orphan_id__": "__$orphan_id__(0)",
  "__META__": Object {
    "MODEL_PERSISTED_FIELD": false,
    "collection": undefined,
    "fields": Object {
      "__$orphan_id__": Object {
        "defaultValue": undefined,
        "referenceDef": false,
      },
      "staff": Object {
        "defaultValue": undefined,
        "referenceDef": Object {
          "model": "organization.Staff",
          "type": 0,
        },
      },
    },
    "id": "__$orphan_id__(0)",
    "type": "organization.Staff.Me",
  },
  "staff": Object {
    "id": "cdb28c900c75",
    "type": "organization.Staff",
  },
}
`;

exports[`collection.fetch should be fetched all Refs Models fetchModelRefs 2`] = `
Object {
  "__META__": Object {
    "MODEL_PERSISTED_FIELD": true,
    "collection": undefined,
    "fields": Object {
      "code": Object {
        "defaultValue": undefined,
        "referenceDef": false,
      },
      "defaultDepartment": Object {
        "defaultValue": undefined,
        "referenceDef": Object {
          "model": "organization.Department",
          "type": 0,
        },
      },
      "departments": Object {
        "defaultValue": undefined,
        "referenceDef": Object {
          "model": "organization.Department",
          "property": undefined,
          "type": 1,
        },
      },
      "id": Object {
        "defaultValue": undefined,
        "referenceDef": false,
      },
      "name": Object {
        "defaultValue": undefined,
        "referenceDef": false,
      },
    },
    "id": "cdb28c900c75",
    "type": "organization.Staff",
  },
  "code": 41738,
  "defaultDepartment": Object {
    "id": "66f58b59c384",
    "type": "organization.Department",
  },
  "departments": Array [
    Object {
      "id": "66f58b59c384",
      "type": "organization.Department",
    },
  ],
  "id": "cdb28c900c75",
  "name": "Principal",
}
`;
